crates/sequencer/src/sequencer.rs

```diff
@@ -564,26 +429,123 @@ where
+                // Receive updates from DA layer worker.
+                l1_data = da_rx.recv() => {
+                    // Stop receiving updates from DA layer until we have caught up.
+                    if missed_da_blocks_count > 0 {
+                        continue;
+                    }
+                    if let Some(l1_data) = l1_data {
+                        (prev_l1_height, last_finalized_block, l1_fee_rate) = l1_data;
+
+                        if last_finalized_height > prev_l1_height {
+                            let skipped_blocks = last_finalized_height - prev_l1_height;
+                            if skipped_blocks > 1 {
+                                // This shouldn't happen. If it does, then we should produce at least 1 block for the blocks in between
+                                warn!(
+                                    "Sequencer is falling behind on L1 blocks by {:?} blocks",
+                                    skipped_blocks
+                                );
+
+                                // Missed DA blocks means that we produce n - 1 empty blocks, 1 per missed DA block.
+                                missed_da_blocks_count = skipped_blocks - 1;
+                            }
+                        }
+
+                        if let Err(e) = self.submit_commitment(last_finalized_height, prev_l1_height).await {
+                            error!("Sequencer error: {}", e);
+                        }
+                    }
+                },
+                // If sequencer is in test mode, it will build a block every time it receives a message
+                // The RPC from which the sender can be called is only registered for test mode. This means
+                // that evey though we check the receiver here, it'll never be "ready" to be consumed unless in test mode.
+                _ = self.l2_force_block_rx.next() => {
+                    if self.config.test_mode {
+                        if let Err(e) = self.produce_l2_block(last_finalized_block.clone(), l1_fee_rate, L2BlockMode::NotEmpty).await {
+                            error!("Sequencer error: {}", e);
+                        }
+                    }
+                },
+                // If sequencer is in production mode, it will build a block every 2 seconds
+                _ = interval.tick() => {
+                    // By default, we produce a non-empty block IFF we were caught up all the way to
+                    // last_finalized_block. If there are missed DA blocks, we start producing
+                    // empty blocks at ~2 second rate, 1 L2 block per respective missed DA block
+                    // until we know we caught up with L1.
+                    let mut l2_block_mode = L2BlockMode::NotEmpty;
+                    let mut da_block = last_finalized_block.clone();
+
+                    if missed_da_blocks_count > 0 {
+                        l2_block_mode = L2BlockMode::Empty;
```

Comment:
I mean if we could account for msised L1 blocks without waiting 2 seconds, it would be much better.

But we're operating on Bitcoin, and this will happen only like once or twice, and it will be a single missed_da_block.

So, if it's hard to implement this can stay as is.